File: midi-script/DrumPad.py
from __future__ import absolute_import

from .Interface import Interface


class DrumPad(Interface):
    @staticmethod
    def serialize_drum_pad(pad):
        if pad is None:
            return None

        pad_id = Interface.save_obj(pad)

        mute = False
        solo = False
        note = None

        try:
            mute = pad.mute
        except:
            pass

        try:
            solo = pad.solo
        except:
            pass

        try:
            note = pad.note
        except:
            pass

        return {
            "id": pad_id,
            "name": pad.name,
            "note": note,
            "mute": mute,
            "solo": solo,
        }

    def __init__(self, c_instance, socket):
        super(DrumPad, self).__init__(c_instance, socket)

    def get_chains(self, ns):
        from .Chain import Chain
        return map(Chain.serialize_chain, ns.chains)

File: src/ns/drum-pad.ts
import { Ableton } from "../index.js";
import { Namespace } from "./index.js";
import { DrumChain, RawDrumChain } from "./drum-chain.js";

export interface GettableProperties {
  chains: RawDrumChain[];
  mute: boolean;
  name: string;
  note: number | null;
  solo: boolean;
}

export interface TransformedProperties {
  chains: DrumChain[];
}

export interface SettableProperties {
  mute: boolean;
  solo: boolean;
}

export interface ObservableProperties {
  chains: RawDrumChain[];
  mute: boolean;
  name: string;
  solo: boolean;
}

export interface RawDrumPad {
  readonly id: string;
  readonly name: string;
  readonly note: number | null;
  readonly mute: boolean;
  readonly solo: boolean;
}

/**
 * Represents a single pad of a Drum Rack device. Reach a pad's own device
 * chain (e.g. its Simpler) via `chains` - a pad usually has 0 or 1 chains.
 */
export class DrumPad extends Namespace<
  GettableProperties,
  TransformedProperties,
  SettableProperties,
  ObservableProperties
> {
  constructor(
    ableton: Ableton,
    public readonly raw: RawDrumPad,
  ) {
    super(ableton, "drum-pad", raw.id);

    this.transformers = {
      chains: (chains) => chains.map((c) => new DrumChain(ableton, c)),
    };

    this.cachedProps = {
      chains: true,
    };
  }

  /** Deletes all chains on this pad (same as clearing a drum rack pad in Live). */
  deleteAllChains() {
    return this.sendCommand("delete_all_chains");
  }
}
